src/cljc/matcher_combinators/matchers.cljc

```diff
@@ -71,9 +71,79 @@
   [pred]
   (core/->PredMatcher pred (str "predicate: " pred)))
 
+#?(:cljs (defn- cljs-uri [expected]
+           (core/->CljsUriEquals expected)))
+
 (defn matcher-for
-  "Returns the type-specific matcher object for an expected value. This is
-  useful for discovery when you want to know which Matcher type is associated
-  to a value."
-  [expected]
-  (core/-matcher-for expected))
+  "Returns the type-specific matcher object for an expected
+  value. This is used internally to support the match-with matcher,
+  and is also useful for discovery when you want to know which Matcher
+  type is associated to a value.
+
+  Adds :matcher-object? metadata to the returned matcher so that
+  other functions can differentiate between matcher objects and
+  objects that happen to implement the Matcher protocol (which should
+  be all other objects)."
+  ([expected]
+   (core/-matcher-for expected))
+  ([expected overrides]
+   (core/-matcher-for expected overrides)))
+
+(declare match-with)
+
+(defn- match-with-values [m overrides]
+  (reduce-kv (fn [m* k v]
+               (assoc m* k (match-with overrides v)))
+             {}
+             m))
+
+(defn- match-with-elements [coll overrides]
+  (reduce (fn [c v]
+            (conj c (match-with overrides v)))
+          (empty coll)
+          coll))
+
+(defn match-with [overrides value]
+  (vary-meta
+   (cond (:match-with? (meta value))
+         value
+
+         (fn? value)
+         value
+
+         ;; TODO: all of the built in matchers are records, but users
+         ;; define matchers by reifying the Matcher protocol, so this
+         ;; would break down. Also, what if a user's domain includes a
+         ;; record with an `:expected` key? Ideally, we should have
+         ;; some other marker to identify a matcher object, and document
+         ;; it in terms of "your custom Matcher implementations must do
+         ;; x in order to particpate in match-with"
+         (and (record? value) (map? (:expected value)))
```

Comment:
Where is `:expected` being assigned?

Reply:
It's part of all of the matcher record implementations.

Reply:
Oh, it is the parameter of all records. Unfortunately we cannot use a namespaced keyword :(
https://clojure.atlassian.net/browse/CLJ-1938